### PR TITLE
interpret `tuple` as a class and `tuple[]` as the empty tuple, enable syntax for anonymous tuples

### PR DESCRIPTION
**Pull request by Maxdamantus**
When the indentation syntax is allowed it is always interpreted as a
tuple:

    type
      Unit = tuple

### CODE REVIEW

**Comment by Maxdamantus** [comment]
Currently, the empty tuple type is kind of interpreted as a type class.
Actual empty tuple types are logical, and can be useful as a primitive unit type in generics, for example, a set type could be defined in terms of ``Table[T, tuple[]]``, or you could have a generic parser type ``Parser[Tok, Output]``, where a ``Parser[string, tuple[]]`` would simply consume input and not result in a meaningful value.

Side note: I think ``void`` in C, C++, Java, etc was largely a mistake and should have used a unit type (which doesn't actually add any overhead, as it would optimally not require any storage).

File: compiler/ast.nim [hunk 1]
```diff
@@ -196,6 +196,7 @@ type
     nkTypeOfExpr,         # type(1+2)
     nkObjectTy,           # object body
     nkTupleTy,            # tuple body
+    nkTupleClassTy,       # tuple type class
```
Comment (Araq) [inline compiler/ast.nim]: You need to update lib/core/macros.nim too when you add a new AST kind. Note that this doesn't mean I'll accept this PR. I don't know yet.

Comment (Maxdamantus) [inline compiler/ast.nim]: I think I did that—though I updated it in the  form of a squash rebase, so it might've gone unnoticed.

**Comment by fowlmouth** [comment]
I don't understand, `tuple` already is a type class which is quite different than `tuple[]` 
```nim
proc foo (some: tuple) =
  echo repr(some)
foo((1,2))
```

**Comment by Maxdamantus** [comment]
On Mon, Mar 02, 2015 at 11:04:08AM -0800, Billingsly Wetherfordshire wrote:
> I don't understand, `tuple` already is a type class which is quite different than `tuple[]` 
> ```nim
> proc foo (some: tuple) =
>   echo repr(some)
> foo((1,2))
> ```
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/Araq/Nim/pull/2239#issuecomment-76783031
They currently mean the same thing. The empty tuple and its type exists,
and in error messages it's described as `tuple[]`, but when you actually
write `tuple[]` it's the same as `tuple`:

  proc semTuple(c: PContext, n: PNode, prev: PType): PType =
    if n.sonsLen == 0: return newConstraint(c, tyTuple)
    ..

**Pull request merged**